File: src/EggsAndHoney.WebApi/Controllers/ResolvedOrdersController.cs
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using AutoMapper;
using EggsAndHoney.Domain.Services;
using EggsAndHoney.WebApi.Filters;
using EggsAndHoney.WebApi.ViewModels;
using Microsoft.AspNetCore.Mvc;

namespace EggsAndHoney.WebApi.Controllers
{
    [Route("api/v1/[controller]")]
    [ValidateModel]
    public class ResolvedOrdersController : Controller
    {
        private readonly IMapper _mapper;
        private readonly IOrderService _orderService;

        public ResolvedOrdersController(IMapper mapper, IOrderService orderService)
        {
            _mapper = mapper;
            _orderService = orderService;
        }

        [HttpGet]
		[ProducesResponseType(typeof(ItemCollectionResponseViewModel<ResolvedOrderViewModel>), 200)]
        public async Task<IActionResult> Get()
        {
            var resolvedOrders = await _orderService.GetResolvedOrders();

            var resolvedOrderViewModels = _mapper.Map<IList<ResolvedOrderViewModel>>(resolvedOrders.OrderByDescending(o => o.DateResolved));
            var itemCollectionResponseViewModel = new ItemCollectionResponseViewModel<ResolvedOrderViewModel>(resolvedOrderViewModels);

            return Ok(itemCollectionResponseViewModel);
        }

        [HttpPost("unresolve")]
		[ProducesResponseType(typeof(OrderViewModel), 200)]
		[ProducesResponseType(400)]
		[ProducesResponseType(404)]
		public async Task<IActionResult> Unresolve([FromBody]ItemIdentifierViewModel itemIdentifier)
        {
			if (itemIdentifier == null)
			{
				return BadRequest("Request body cannot be empty!");
			}

            if (!await _orderService.ResolvedOrderExists(itemIdentifier.Id))
            {
                return NotFound();
            }

            var unresolvedOrder = await _orderService.UnresolveOrder(itemIdentifier.Id);
            var unresolvedOrderViewModel = _mapper.Map<OrderViewModel>(unresolvedOrder);

            return Ok(unresolvedOrderViewModel);
        }
    }
}

File: src/EggsAndHoney.WebApi.FSharp/Controllers/ResolvedOrdersController.fs
namespace EggsAndHoney.WebApi.FSharp.Controllers

open System
open Microsoft.AspNetCore.Mvc
open EggsAndHoney.WebApi.FSharp.ViewModels
open EggsAndHoney.WebApi.FSharp.Mapping
open EggsAndHoney.Domain.Services
open EggsAndHoney.Domain.Models

// TODO: Add model validation attribute in order to get the rest of the tests to pass
[<Route("api/v1/[controller]")>]
type ResolvedOrdersController (orderService: IOrderService) =
    inherit Controller()

    [<HttpGet>]
    [<ProducesResponseType(typeof<ItemCollectionResponseViewModel<OrderViewModel>>, 200)>]
    member this.Get() =
        async {
            let! resolvedOrders = orderService.GetResolvedOrders() |> Async.AwaitTask
            return this.Ok(
                resolvedOrders |>
                Seq.sortByDescending (fun o -> o.DateResolved) |>
                toResolvedOrderCollectionResponseViewModel)
        }

    [<HttpPost("unresolve")>]
    [<ProducesResponseType(typeof<OrderViewModel>, 200)>]
    [<ProducesResponseType(400)>]
    [<ProducesResponseType(404)>]
    member this.Unresolve([<FromBody>] itemIdentifier) : Async<IActionResult> =
        async {
            let! resolvedOrderExists = orderService.ResolvedOrderExists(itemIdentifier.id) |> Async.AwaitTask
            match resolvedOrderExists with
            | false -> return this.NotFound() :> _
            | _ ->
                let! unresolvedOrder = orderService.UnresolveOrder(itemIdentifier.id) |> Async.AwaitTask
                return this.Ok(unresolvedOrder |> toOrderViewModel) :> _
        }

